### ISSUE TITLE
[python] Detect class attributes

### ISSUE THREAD

**Issue by balta2ar** [universal-ctags/ctags#1108]
Consider this code:
``` python
class TestClass(object):
    def __init__(self):
        self.public_member_init = None
        self._private_member_init = None

    def public_method(self):
        pass

    def _private_method(self):
        pass

def main():
    instance = TestClass()
    instance._private_member_init = 1
```

The output:
```
$ ctags --python-kinds=+zl -o- test.py
TestClass       test.py /^class TestClass(object):$/;"  c
__init__        test.py /^    def __init__(self):$/;"   m       class:TestClass
_private_method test.py /^    def _private_method(self):$/;"    m       class:TestClass
instance        test.py /^    instance = TestClass()$/;"        l       function:main   file:
main    test.py /^def main():$/;"       f
public_method   test.py /^    def public_method(self):$/;"      m       class:TestClass
self    test.py /^    def __init__(self):$/;"   z       member:TestClass.__init__       file:
self    test.py /^    def _private_method(self):$/;"    z       member:TestClass._private_method        file:
self    test.py /^    def public_method(self):$/;"      z       member:TestClass.public_method  file:
```

I wish there were `public_member_init` and `_private_member_init` in the tags. Is this technically possible in Python? If possible, what might be the effort required to implement it? I wonder if it's a realistic feature request of something way too complicated. Thank you!

**Comment by b4n** [comment]
It would at least be awfully tricky.  The problem here is that assigning members to `self` just happens to work because `self` refers to the class instance, but self is just the first argument of `__init__()`, not any kind of keyword (it's customary to use this name, but not required).

So the parser would have to:
* understand this is a non-static *class method* (e.g. not wrapped with dynamic stuff like `@classmethod`.  Accurately finding this out is impossible short of compiling the Python code.
* get the instance method argument (e.g., first one)
* find assignations to member of that instance, probably making sure not to emit it twice.

A hacky support of `self.<identifier> = ` inside `__init__()` alone would probably be simpler, but it wouldn't be technically correct.  But maybe it'd work good enough most of the time, not sure.

**Comment by balta2ar** [comment]
I totally agree that making it fully technically correct in all edge cases is a tricky task. However, at least in the code base that I'm working with and in my code style, it's a good practice to initialize class attributes in __init__ method and it's always encouraged to do so. With this in mind, I think it might be a good idea to follow Pareto principle: with 20% of effort make it work in 80% cases.

**Comment by masatake** [comment]
With --fields=+e, you can know the both start line and end line of `__init__`.
You can cut down the lines within the range with sed.
Then pick up the pattern `self\.(.*) =` with sed, again.
Lines that can be appended to tags file can be genearted.

You may want to do it semi-automatically inside ctags...
What kind of user interface can we provide?

```
--regex-python=/self\.(.*)[ \t]*=/\1/i,instanceVariable,instance variables/{area=__init__}
```

Interesting to me. The ctags main part and the new python parser provides enough infrastructure for 
implementing this command line.  However, I will not work this topic now. I should make a release.
Instead you can hack as you want. What I can do now it just giving hints (if you want :-P).

https://github.com/universal-ctags/ctags/issues/80 is a bit related to this issue.

**Comment by masatake** [comment]
I'm working on introducing a symbol table that is extended from cork queue.
This can be used as a building block to implement the heuristic approaches we discussed here.

As a code reader, I need tags for class attributes. If we can capture them, reading python codes will be much easier.

The comment I made in 2016 is still interesting for me. Running matching extra regex patterns after running its base parser may be powerful and will be slow.

```
--kinddef-Python=i,instanceVariable,instance variables
--pass2-regex-python=/self\.(.*)[ \t]*=/\1/i/{area=member:__init__}
```

This will be extremely powerful those who want to capture reference tags easily.

In 2016, I was talking about "releasing," and I have not yet made a release.

**Comment by masatake** [comment]
```Python
class BASE(object):
    def __init__(self):
        self.b = 1

class TestClass(BASE):
    def __init__(self):
        self.public_member_init = None
        self._private_member_init = None
        self.b = 2
    def m0(self):
        self.a = 1
    def m1(self):
        self.a = 3
```

Assumption: we have Level1 symbol table.  A parser can look up a name already defined in the current input file. Only names defined ahead of the current input line are in the L1 symtab.

Heuristic 1. We can capture foo in `self.foo = `in __init__ method in a class as a class attribute. `public_member_init`  and `_private_member_init` are captured, and we are happy about these. How about `self.b` TestClass class? `self.b` are in both BASE class and TestCase class. Only `b` in `self.b` in BASE should be captured? 
If we are talking about C++ or Java, only `b` in the base class should be captured. However, should we apply this rule to dynamically typed language? In C++ or Java, we may want on the base class because, I think, people want to know the type of `b`.
Actually field definition can be done only one place. In a dynamically typed language, tagging twice both in the base class and its descendant is acceptable I think as far as scope field and inherits fields are filled correctly in the parser.

How about `a` in `self.a`?
Heuristic 2. capture `a` appeared in the first time. It measing capturing `a` in `self.a` in `m0` method. Here comes the L1 symbol table.  When the python parser finds `self.a` in a member function (method), the parser looks up the name in the L1  symtab. If an entry is returned, the instance variable is defined already — no need to capture the same one.

**Comment by masatake** [comment]
Now Universal-ctags has the Level1 symbol table called 'symtab'.

**Comment by thouger** [comment]
still wait a answer,and the regex use `/self\.(.*)[ \t]*=[ \t]*/\1/i/` be better.now has the answer to solve `self.config = config` this ?

**Comment by masatake** [comment]
I have not implemented `{area=__init__}` flag yet. The flag may be just ignored.

**Comment by thouger** [comment]
> I have not implemented `{area=__init__}` flag yet. The flag may be just ignored.

Thank you very much for your reply and your contribution to this project.universal-ctags is useful,There's no problem at present
